File: trees/cppRealizations/validateBinarySearchTree.cpp
#include "btf.h"
#include <climits>


class Solution {
public:
    bool dfs(TreeNode* root, int min, int max){
        if(root == nullptr){
            return true;
        }
        if(!(root->val > min && root->val < max)){
            return false;
        }
        return dfs(root->left, min, root->val) && dfs(root->right, root->val, max);
    }

    bool isValidBST(TreeNode* root) {
        return dfs(root, INT_MIN, INT_MAX);
    }
};


int main(){
    TreeNode* root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);
    cout << "Input:" << endl;
    btf(root);
    cout << "Sulution:" << endl;
    cout << Solution().isValidBST(root) << endl;
    return 0;
}

File: trees/validateBinarySearchTree.py
from trees.btf import *


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def valid(node, left, right):
            if not node:
                return True
            if not (node.val < right and node.val > left):
                return False
            return valid(node.left, left, node.val) and valid(node.right, node.val, right)
        
        return valid(root, -float('inf'), float('inf'))


if __name__ == "__main__":
    root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)
    print("Input:")
    btf(root)
    print("Sulution:")
    print(Solution().isValidBST(root))